File: web/src/feature/session/type/content-block.ts
import type { Timestamp } from '@bufbuild/protobuf/wkt'
import type { UsageMetadata } from '@/gen/grpcstub/collector/v1/collector_pb'

// Content block types for assistant messages
export interface TextContentBlock {
  type: 'text'
  text: string
}

export interface ToolUseContentBlock {
  type: 'tool_use'
  id: string
  name: string
  input: Record<string, unknown>
}

export interface ToolResultContentBlock {
  type: 'tool_result'
  tool_use_id: string
  content: string
  is_error?: boolean
}

export type AssistantContentBlock = TextContentBlock | ToolUseContentBlock
export type ContentBlock = TextContentBlock | ToolUseContentBlock | ToolResultContentBlock

// Parsed message structure for rendering
export interface ParsedMessage {
  uuid: string
  type: 'user' | 'assistant' | 'system' | 'tool_result'
  timestamp?: Timestamp
  isMeta: boolean
  isSidechain: boolean
  usage?: UsageMetadata
  content: ContentBlock[]
  // For tool_result messages
  toolName?: string
  parentToolUseId?: string
}

// Tool call with linked result for panel display
export interface LinkedToolCall {
  toolUse: ToolUseContentBlock
  toolResult?: ToolResultContentBlock
  sourceMessageUuid: string
  timestamp?: Timestamp
}

File: shared/domain/content_block.go
package domain

// ContentBlockType represents the type discriminator for content blocks.
type ContentBlockType string

const (
	ContentBlockTypeText       ContentBlockType = "text"
	ContentBlockTypeToolUse    ContentBlockType = "tool_use"
	ContentBlockTypeToolResult ContentBlockType = "tool_result"
)

// ContentBlock is the interface implemented by all content block types.
type ContentBlock interface {
	BlockType() ContentBlockType
}

// TextContentBlock represents a text content block.
type TextContentBlock struct {
	Type ContentBlockType `json:"type"`
	Text string           `json:"text"`
}

// BlockType implements ContentBlock interface.
func (b *TextContentBlock) BlockType() ContentBlockType { return ContentBlockTypeText }

// ToolUseContentBlock represents a tool use content block.
type ToolUseContentBlock struct {
	Type  ContentBlockType `json:"type"`
	ID    string           `json:"id"`
	Name  string           `json:"name"`
	Input map[string]any   `json:"input"`
}

// BlockType implements ContentBlock interface.
func (b *ToolUseContentBlock) BlockType() ContentBlockType { return ContentBlockTypeToolUse }

// ToolResultContentBlock represents a tool result content block.
type ToolResultContentBlock struct {
	Type      ContentBlockType `json:"type"`
	ToolUseID string           `json:"tool_use_id"`
	Content   string           `json:"content"`
	IsError   bool             `json:"is_error,omitempty"`
}

// BlockType implements ContentBlock interface.
func (b *ToolResultContentBlock) BlockType() ContentBlockType { return ContentBlockTypeToolResult }
